src/IO/AsynchronousReadBufferFromFileDescriptor.cpp

```diff
@@ -153,46 +183,48 @@ off_t AsynchronousReadBufferFromFileDescriptor::seek(off_t offset, int whence)
     if (new_pos + (working_buffer.end() - pos) == file_offset_of_buffer_end)
         return new_pos;
 
-    if (file_offset_of_buffer_end - working_buffer.size() <= static_cast<size_t>(new_pos)
-        && new_pos <= file_offset_of_buffer_end)
+    while (true)
     {
-        /// Position is still inside the buffer.
-        /// Probably it is at the end of the buffer - then we will load data on the following 'next' call.
+        if (file_offset_of_buffer_end - working_buffer.size() <= new_pos && new_pos <= file_offset_of_buffer_end)
+        {
+            /// Position is still inside the buffer.
+            /// Probably it is at the end of the buffer - then we will load data on the following 'next' call.
 
-        pos = working_buffer.end() - file_offset_of_buffer_end + new_pos;
-        assert(pos >= working_buffer.begin());
-        assert(pos <= working_buffer.end());
+            pos = working_buffer.end() - file_offset_of_buffer_end + new_pos;
+            assert(pos >= working_buffer.begin());
+            assert(pos <= working_buffer.end());
 
-        return new_pos;
-    }
-    else
-    {
-        if (prefetch_future.valid())
+            return new_pos;
+        }
+        else if (prefetch_future.valid())
         {
-            //std::cerr << "Ignoring prefetched data" << "\n";
-            prefetch_future.wait();
-            prefetch_future = {};
+            /// Read from prefetch buffer and recheck if the new position is valid inside.
+
+            if (nextImpl())
+                continue;
         }
 
-        /// Position is out of the buffer, we need to do real seek.
-        off_t seek_pos = required_alignment > 1
-            ? new_pos / required_alignment * required_alignment
-            : new_pos;
+        break;
+    }
 
-        off_t offset_after_seek_pos = new_pos - seek_pos;
+    assert(!prefetch_future.valid());
 
-        /// First reset the buffer so the next read will fetch new data to the buffer.
-        resetWorkingBuffer();
+    /// Position is out of the buffer, we need to do real seek.
+    off_t seek_pos = required_alignment > 1
+        ? new_pos / required_alignment * required_alignment
+        : new_pos;
 
-        /// Just update the info about the next position in file.
+    /// First reset the buffer so the next read will fetch new data to the buffer.
+    resetWorkingBuffer();
 
-        file_offset_of_buffer_end = seek_pos;
+    /// Just update the info about the next position in file.
 
-        if (offset_after_seek_pos > 0)
-            ignore(offset_after_seek_pos);
+    file_offset_of_buffer_end = seek_pos;
+    bytes_to_ignore = new_pos - seek_pos;
 
-        return seek_pos;
-    }
+    assert(bytes_to_ignore < internal_buffer.size());
```

Comment:
@kssenii @amosbird 

It's possible that we mistakenly resized `internal_buffer` to some small value because we were reading near the end of the file. For example, the `internal_buffer.size()` can be 123 here.

Then we seek back to read another part of the file. Say, at offset 4396. But due to alignment requirements, we have to seek to offset 4096 and ignore 300 bytes. But 300 is larger than 123.

Reply:
https://github.com/ClickHouse/ClickHouse/pull/33653/commits/04b7a34870c8f5dba7f20d5d7f8ffb338175e740